File: ios/drappula/Consent/ConsentManagerIOS.swift
import FirebaseAnalytics
import FirebaseCrashlytics
import FirebasePerformance
import Foundation

struct ConsentStateIOS {
    let analytics: Bool
    let crashReporting: Bool
}

class ConsentManagerIOS {
    static let shared = ConsentManagerIOS()

    private let defaults = UserDefaults.standard

    private enum Keys {
        static let analytics = "consent_analytics"
        static let crashReporting = "consent_crash_reporting"
        static let hasResponded = "consent_has_responded"
    }

    var consentState: ConsentStateIOS {
        ConsentStateIOS(
            analytics: defaults.bool(forKey: Keys.analytics),
            crashReporting: defaults.bool(forKey: Keys.crashReporting)
        )
    }

    func updateConsent(_ state: ConsentStateIOS) {
        defaults.set(state.analytics, forKey: Keys.analytics)
        defaults.set(state.crashReporting, forKey: Keys.crashReporting)
        defaults.set(true, forKey: Keys.hasResponded)
        applyToFirebase(state)
    }

    func hasUserResponded() -> Bool {
        defaults.bool(forKey: Keys.hasResponded)
    }

    func applyToFirebase(_ state: ConsentStateIOS) {
        Analytics.setAnalyticsCollectionEnabled(state.analytics)
        Crashlytics.crashlytics().setCrashlyticsCollectionEnabled(state.crashReporting)
        Performance.sharedInstance().isDataCollectionEnabled = state.analytics
        Performance.sharedInstance().isInstrumentationEnabled = state.analytics

        Analytics.setConsent([
            .analyticsStorage: state.analytics ? .granted : .denied,
            .adStorage: .denied,
            .adUserData: .denied,
            .adPersonalization: .denied
        ])
    }
}

File: shared/src/iosMain/kotlin/com/hellocuriosity/drappula/consent/ConsentManager.ios.kt
package com.hellocuriosity.drappula.consent

import platform.Foundation.NSUserDefaults

actual class ConsentManager {
    private val defaults = NSUserDefaults.standardUserDefaults

    actual fun getConsentState(): ConsentState =
        ConsentState(
            analytics = defaults.boolForKey(ANALYTICS),
            crashReporting = defaults.boolForKey(CRASH_REPORTING),
        )

    actual fun updateConsent(state: ConsentState) {
        defaults.setBool(state.analytics, forKey = ANALYTICS)
        defaults.setBool(state.crashReporting, forKey = CRASH_REPORTING)
        defaults.setBool(true, forKey = HAS_RESPONDED)
    }

    actual fun hasUserResponded(): Boolean = defaults.boolForKey(HAS_RESPONDED)

    companion object {
        private const val ANALYTICS = "ANALYTICS"
        private const val CRASH_REPORTING = "CRASH_REPORTING"
        private const val HAS_RESPONDED = "HAS_RESPONDED"
    }
}
